File: backend/asterion_api/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class HealthResponse(BaseModel):
    status: Literal["ok", "degraded"]
    app: str
    uptime_seconds: float
    database: dict[str, Any]
    privacy: dict[str, Any]


class ModelInfo(BaseModel):
    name: str
    modified_at: str | None = None
    size: int | None = None
    digest: str | None = None


class ModelsResponse(BaseModel):
    models: list[ModelInfo]
    privacy_level: Literal["local"] = "local"


class ChatRequest(BaseModel):
    message: str = Field(min_length=1, max_length=64_000)
    room_id: str = Field(default="default", min_length=1, max_length=256)
    conversation_id: str | None = None
    model: str | None = None
    files_attached: bool = False
    memory_enabled: bool = False
    web_access: bool = False


class ChatResponse(BaseModel):
    conversation_id: str
    room_id: str
    model: str
    response: str
    latency_ms: float
    privacy_level: Literal["local"] = "local"
    ts: datetime


class PrivacyItem(BaseModel):
    what: str
    destination: str
    risk: Literal["green", "yellow", "red"]


class PrivacyReport(BaseModel):
    level: Literal["green", "yellow", "red"]
    items: list[PrivacyItem]


class PrivacyAnalyzeRequest(BaseModel):
    model_type: Literal["local", "api"] = "local"
    files_attached: bool = False
    memory_enabled: bool = False
    web_access: bool = False


class HardwareProfile(BaseModel):
    vram_gb: float = 0
    ram_gb: float | None = None
    gpu_name: str | None = None


class ModelSelectRequest(BaseModel):
    task_description: str
    hw_profile: HardwareProfile


class ModelSelection(BaseModel):
    model: str
    mode: Literal["local", "api"]
    reason: str


class MemoryCreateRequest(BaseModel):
    room_id: str = Field(min_length=1, max_length=256)
    content: str = Field(min_length=1, max_length=64_000)
    source: str = Field(default="manual", min_length=1, max_length=256)
    expires_at: datetime | None = None
    privacy: PrivacyAnalyzeRequest = Field(default_factory=PrivacyAnalyzeRequest)


class MemoryUpdateRequest(BaseModel):
    content: str | None = Field(default=None, min_length=1, max_length=64_000)
    source: str | None = Field(default=None, min_length=1, max_length=256)
    expires_at: datetime | None = None


class MemoryRecord(BaseModel):
    id: str
    room_id: str
    content: str
    source: str
    created_at: datetime
    expires_at: datetime | None = None
    privacy: PrivacyReport | None = None


class RagIndexRequest(BaseModel):
    file_path: str
    room_id: str = "default"


class RagSearchRequest(BaseModel):
    query: str = Field(min_length=1, max_length=16_000)
    room_id: str = "default"
    limit: int = Field(default=8, ge=1, le=50)


class RagChunk(BaseModel):
    id: str
    room_id: str
    content: str
    source: str
    score: float = 0


class DeepResearchRequest(BaseModel):
    query: str = Field(min_length=1, max_length=16_000)
    max_subtasks: int = Field(default=5, ge=3, le=5)
    web_access: bool = True


class ResearchResult(BaseModel):
    subtask: str
    title: str
    url: str | None = None
    snippet: str | None = None


class DeepResearchResponse(BaseModel):
    query: str
    subtasks: list[str]
    results: list[ResearchResult]
    privacy: PrivacyReport


class ContradictionRequest(BaseModel):
    claims: list[str] = Field(min_length=2)
    threshold: float = Field(default=0.85, ge=0, le=1)


class ContradictionMatch(BaseModel):
    left: str
    right: str
    similarity: float
    sentiment_left: str
    sentiment_right: str


class AgentPermissions(BaseModel):
    allowed_folders: list[str] = Field(default_factory=list)
    network: bool = False
    shell: bool = False


class RuntimeSkillManifest(BaseModel):
    id: str
    name: str
    category: str
    description: str
    privacy_level: Literal["local", "hybrid", "external"]
    inputs: list[str] = Field(default_factory=list)
    outputs: list[str] = Field(default_factory=list)
    tools: list[str] = Field(default_factory=list)
    guardrails: list[str] = Field(default_factory=list)


class AgentManifest(BaseModel):
    id: str
    name: str
    role: str
    description: str
    privacy_level: Literal["local", "hybrid", "external"]
    default_model: str
    skills: list[str] = Field(default_factory=list)
    permissions: AgentPermissions = Field(default_factory=AgentPermissions)
    system_prompt: str
    escalation_policy: str


class AgentCatalog(BaseModel):
    agents: list[AgentManifest]
    skills: list[RuntimeSkillManifest]


class AgentPlan(BaseModel):
    steps: list[str]
    required_permissions: list[str]
    estimated_tokens: int


class AgentRunCodeRequest(BaseModel):
    code: str = Field(min_length=1, max_length=32_000)
    permissions: AgentPermissions = Field(default_factory=AgentPermissions)


class ComfyGenerateRequest(BaseModel):
    prompt: str = Field(min_length=1, max_length=16_000)
    recipe: dict[str, Any] = Field(default_factory=dict)


class WorkflowRunRequest(BaseModel):
    workflow: dict[str, Any]


class WorkflowConfirmRequest(BaseModel):
    run_id: str
    approved: bool = True
    payload: dict[str, Any] = Field(default_factory=dict)


class PluginManifest(BaseModel):
    name: str
    trust_level: Literal["verified", "local-only", "network", "file", "shell", "danger"]
    path: str
    description: str | None = None


class WorkflowRunStatus(BaseModel):
    id: str
    status: str
    workflow: dict[str, Any]
    results: list[dict[str, Any]]
    created_at: str
    updated_at: str

File: crates/core/src/schemas.rs
use serde::{Deserialize, Serialize};

/// Hardware profile describing available GPU/CPU resources.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HardwareProfile {
    pub vram_gb: f64,
    pub ram_gb: Option<f64>,
}

/// Result of model selection routing.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelSelection {
    pub model: String,
    pub mode: String,
    pub reason: String,
}

/// Entry in the local model catalog.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelEntry {
    pub model: String,
    pub required_vram_gb: f64,
    pub ram_gb: f64,
    pub tags: Vec<String>,
    pub reason: String,
}

/// Benchmark result for a model.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BenchmarkResult {
    pub model: String,
    pub tokens_per_second: f64,
    pub latency_ms: f64,
    pub vram_gb: f64,
    pub cache_hit: bool,
    pub ts: String,
}

/// Plugin manifest loaded from a plugin directory.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PluginManifest {
    pub name: String,
    pub trust_level: String,
    pub path: String,
    pub description: Option<String>,
}

// ── Privacy ─────────────────────────────────────────────────────────────────

/// Single privacy item in a report.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PrivacyItem {
    pub what: String,
    pub destination: String,
    pub risk: String,
}

/// Full privacy analysis report.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PrivacyReport {
    pub level: String,
    pub items: Vec<PrivacyItem>,
}

/// Input parameters for privacy analysis.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PrivacyAnalyzeRequest {
    pub model_type: String,
    pub files_attached: bool,
    pub memory_enabled: bool,
    pub web_access: bool,
    pub prompt: Option<String>,
}

/// Runtime skill manifest — mirrors Python RuntimeSkillManifest.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RuntimeSkillManifest {
    pub id: String,
    pub name: String,
    #[serde(default = "default_version")]
    pub version: String,
    #[serde(default = "default_owner")]
    pub owner: String,
    pub category: String,
    pub description: String,
    pub privacy_level: String,
    #[serde(default)]
    pub triggers: Vec<String>,
    #[serde(default)]
    pub inputs: Vec<String>,
    #[serde(default)]
    pub outputs: Vec<String>,
    #[serde(default)]
    pub tools: Vec<String>,
    #[serde(default)]
    pub guardrails: Vec<String>,
    #[serde(default)]
    pub requires_consent: Vec<String>,
    #[serde(default)]
    pub failure_modes: Vec<String>,
    #[serde(default)]
    pub acceptance_checks: Vec<String>,
}

/// Agent manifest — mirrors Python AgentManifest.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentManifest {
    pub id: String,
    pub name: String,
    #[serde(default = "default_version")]
    pub version: String,
    pub role: String,
    pub description: String,
    pub privacy_level: String,
    pub default_model: String,
    #[serde(default)]
    pub triggers: Vec<String>,
    #[serde(default)]
    pub skills: Vec<String>,
    #[serde(default)]
    pub permissions: AgentPermissions,
    #[serde(default)]
    pub lifecycle: Vec<String>,
    #[serde(default)]
    pub outputs: Vec<String>,
    #[serde(default)]
    pub handoff_targets: Vec<String>,
    #[serde(default)]
    pub acceptance_checks: Vec<String>,
    pub system_prompt: String,
    pub escalation_policy: String,
}

/// Full agent catalog.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentCatalog {
    pub agents: Vec<AgentManifest>,
    pub skills: Vec<RuntimeSkillManifest>,
}

fn default_version() -> String {
    "0.1.0".into()
}

fn default_owner() -> String {
    "asterion".into()
}

/// Result of catalog validation.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ValidationResult {
    pub ok: bool,
    pub agents_count: usize,
    pub skills_count: usize,
    pub errors: Vec<String>,
    pub warnings: Vec<String>,
}

/// A single research result from web search.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResearchResult {
    pub subtask: String,
    pub title: String,
    pub url: Option<String>,
    pub snippet: Option<String>,
}

/// A single chunk from RAG indexing with relevance score.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RagChunk {
    pub id: String,
    pub room_id: String,
    pub content: String,
    pub source: String,
    pub score: f64,
}

/// Deep research response with query decomposition + results.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeepResearchResponse {
    pub query: String,
    pub subtasks: Vec<String>,
    pub results: Vec<ResearchResult>,
    pub privacy: Option<crate::schemas::PrivacyReport>,
}

/// Agent permissions for sandbox execution.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentPermissions {
    #[serde(default)]
    pub allowed_folders: Vec<String>,
    #[serde(default)]
    pub network: bool,
    #[serde(default)]
    pub shell: bool,
}

/// Agent plan generated by the task simulator.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentPlan {
    pub steps: Vec<String>,
    pub required_permissions: Vec<String>,
    pub estimated_tokens: u64,
}

/// Memory record stored in the ledger.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryRecord {
    pub id: String,
    pub room_id: String,
    pub content: String,
    pub source: String,
    pub created_at: String,
    pub expires_at: Option<String>,
    pub privacy: Option<PrivacyReport>,
}

/// Request to create a new memory entry.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryCreateRequest {
    pub room_id: String,
    pub content: String,
    pub source: Option<String>,
    pub expires_at: Option<String>,
    pub privacy: Option<PrivacyAnalyzeRequest>,
}

/// A contradiction match between two claims.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContradictionMatch {
    pub left: String,
    pub right: String,
    pub similarity: f64,
    pub sentiment_left: String,
    pub sentiment_right: String,
}

// ── Benchmark ────────────────────────────────────────────────────────────────

/// Single benchmark run result for a model.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BenchmarkRunResult {
    pub model: String,
    pub runs: usize,
    pub avg_tokens_per_second: f64,
    pub avg_time_to_first_token_ms: f64,
    pub avg_total_time_ms: f64,
    pub min_tps: f64,
    pub max_tps: f64,
    pub stddev_tps: f64,
    pub vram_estimate_gb: f64,
    pub privacy_level: String,
    pub error: Option<String>,
    pub cached_at: Option<f64>,
}
